### PR TITLE
[tc] Add count by value support on client side

### PR DESCRIPTION
**Pull request by LeoLeo718**
[tc] Add count by value support on client side

## Problem Statement
Currently, Venice thin client lacks the ability to perform value-based counting and aggregation operations on the client side. This limitation means that clients need to fetch all data and perform aggregations manually, which is inefficient and can lead to unnecessary data transfer and processing overhead.

We need to implement a countGroupByValue functionality that allows clients to:
- Count occurrences of each distinct value in specified fields
- Return the top K most frequent values
- Support multiple field aggregation in a single request

## Solution
Implemented a new countGroupByValue operation in the Venice thin client using pure client-side computation:

### Key Features:
- **Simple Field Support**: Handles primitive field types (String, Integer, etc.)
- **Top-K Results**: Returns the most frequent values up to the specified K limit
- **Null Handling**: Properly handles null values
- **Utf8 Conversion**: Automatically converts Avro Utf8 values to String
- **Pure Client-Side**: Uses project() to fetch only required fields, then performs aggregation locally
- **Multi-Field Support**: Can aggregate multiple fields in a single request

### Implementation Details:
- **AvroComputeAggregationRequestBuilder**: Builds aggregation requests with validation:
  - Validates topK parameter (must be positive)
  - Validates field names (non-null, non-empty, must exist in schema)
  - Projects required fields for client-side processing
- **AvroComputeAggregationResponse**: Processes results:
  - Aggregates counts across all compute results
  - Handles Utf8 to String conversion
  - Sorts results by count and returns top K entries
  - Supports generic type casting with proper error handling

## Code changes
- Added new code behind existing interfaces. No new configs needed.
- Introduced new validation for input parameters with clear error messages:
  - Validates topK must be positive
  - Validates field names are non-null and non-empty
  - Validates fields exist in schema
- Added type casting error handling for generic operations

## Concurrency-Specific Checks
- Code has no race conditions or thread safety issues.
- Each request creates its own response object - no shared state
- Result aggregation happens in request-scoped maps
- No blocking calls or synchronization needed - pure computation
- Thread-safe by design - immutable request builders and response objects

### Unit tests added:
- **AvroComputeAggregationRequestBuilderTest**: Comprehensive validation tests
  - Valid configurations with single and multiple fields
  - Invalid topK values (0, negative)
  - Null and empty field validation
  - Non-existent field detection
- **AvroComputeAggregationResponseTest**: Result processing tests
  - Simple field aggregation
  - Null value handling
  - Utf8 value conversion
  - Top-K limiting and sorting
  - Type casting error scenarios
  - Mixed data types handling

### Integration tests added:
- **ReadComputeValidationTest.testCountGroupByValueAggregation**: End-to-end test
  - Creates store with simple string fields
  - Populates with test data including duplicates
  - Verifies count aggregation on single fields
  - Verifies count aggregation on multiple fields
  - Tests with various data patterns and sizes

### Test Coverage:
- Input validation: 100% coverage of error cases
- Response processing: All field types and edge cases covered
- Integration: Full workflow from request to response validated
- Type casting: Proper error handling for type mismatches

## Does this PR introduce any user-facing or breaking changes?
No. This is a new feature that doesn't affect existing functionality.
- All changes are backward compatible
- New functionality is opt-in through the new countGroupByValue API
- Existing client operations remain unchanged
- No changes to wire protocol or storage format

### CODE REVIEW

**Comment by LeoLeo718** [comment]
@FelixGV @haoxu07 Hi, guys, I already implemented the VENG-11952 and relational tests, could you help me to review the code? Thank you very much!

File: clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilderTest.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+      when(storeClient.compute()).thenReturn(newComputeRequestBuilder);
+      AvroComputeAggregationRequestBuilder<String> newBuilder =
+          new AvroComputeAggregationRequestBuilder<>(storeClient, schemaReader);
+
+      newBuilder.countGroupByValue(TOP_K, FIELD_1).execute(keySet);
+    }
+    long endTime = System.nanoTime();
+    long duration = TimeUnit.NANOSECONDS.toMillis(endTime - startTime);
+
+    assertTrue(duration < 1000, "Performance test should complete within 1 second");
```
Comment (haoxu07) [inline clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilderTest.java]: Performance test should not be done here as unit test, as test environment changes. Your CI run does not consistently pass, and that is the reason we will need to some performance test certification in ei certification cluster.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    // topK must bigger than 0
+    if (topK <= 0) {
+      throw new VeniceClientException("TopK must be positive");
+    }
+    // field name must not be empty
+    if (fieldNames == null || fieldNames.length == 0) {
+      throw new VeniceClientException("fieldNames cannot be null or empty");
+    }
+
+    // Validate fields exist in schema and are of supported types
```
Comment (haoxu07) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java]: Seems you did not check the type but only check the field is null or not?

**Review by haoxu07 — COMMENTED** [review]
Thanks for preparing this! It seems you add some performance test here, it is not practical to test against performance in unit test. Maybe you could add some integration test, you could check some examples like this: `ReadComputeValidationTest` to see if you can easily do some integration test there.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java (hunk 2 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java]: Thank you for advice! I implemented type check in the newest version!

File: clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilderTest.java (hunk 1 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilderTest.java]: All right! I deleted in updated version!

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      String key = (value == null) ? null : value.toString();
+      valueToCount.merge(key, 1, Integer::sum);
+    }
+
+    // Sort by count in descending order
+    Map<String, Integer> sortedMap = new LinkedHashMap<>();
+    valueToCount.entrySet()
+        .stream()
+        .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
+        .limit(fieldTopKMap.getOrDefault(field, Integer.MAX_VALUE))
```
Comment (haoxu07) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java]: Maybe we could do quick check field exist or not at the beginning of this function to obtain topK. If not exist, just return empty map.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        .forEach(entry -> sortedMap.put(entry.getKey(), entry.getValue()));
+
+    return sortedMap;
```
Comment (haoxu07) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java:51]: Here you just returned a Map of <String, Integer>, user normally expecting a certain type as interface provided:  `<T> Map<T, Integer> getValueToCount(String fieldName);`, you could explore `AvroGenericDeserializer` to see if we could deserialize by requested type. Here we only need to deserialize topK unique values.

**Review by haoxu07 — COMMENTED** [review]
Need one more pass, left 2 comments.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java (hunk 3 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java]: Got it!

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java (hunk 4 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java:51]: I've implemented the field existence check and generic return type, but found that **ComputeGenericRecord.get(field)** returns already deserialized Objects, not byte[]. Do you mean using **AvroGenericDeserializer** for type casting from Object, or expecting data stored as byte[] that needs actual deserialization?

Comment (haoxu07) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java:51]: Maybe you can try: (T) value = record.get(field);

File: clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AvroComputeAggregationResponseTest.java [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+  private Map<String, Integer> fieldTopKMap;
+
+  @BeforeMethod
+  public void setUp() {
+    computeResults = new HashMap<>();
+    fieldTopKMap = new HashMap<>();
+  }
+
+  @Test(description = "Should correctly count simple field values")
+  public void testSimpleFieldCounting() {
```
Comment (haoxu07) [inline clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AvroComputeAggregationResponseTest.java]: Regarding this test class, seems all test functions following these pattern
1. prepare compute results 
2. Provide field name and topK value
3. get response 
4. Assertion
Maybe you can try to write a shared function like:
```
void runTest(<prepare data>, fieldName, topK, <asserting logic>) {
//call <prepare data>
// Construct fieldTopKMap
// call <asserting logic>
}
```
The idea here is to minimize writing redudant code. Besides, you probably need to prepare  fieldTopKMap with multiple fields to see if it works well.

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/storagenode/ReadComputeValidationTest.java [hunk 6]
```diff
@@ -487,4 +488,130 @@ private static List<Float> generateRandomFloatList(int listSize) {
...
+        Assert.assertNotNull(jobTypeCounts);
+
+        // Expected counts: full-time=3, part-time=3
+        Assert.assertEquals(jobTypeCounts.size(), 2, "Should have exactly 2 jobType values");
+        Assert.assertEquals(jobTypeCounts.get("full-time"), Integer.valueOf(3), "full-time count should be 3");
+        Assert.assertEquals(jobTypeCounts.get("part-time"), Integer.valueOf(3), "part-time count should be 3");
+
+        // Test countGroupByValue on location field, topK=2
+        com.linkedin.venice.client.store.ComputeAggregationResponse locationAggResponse =
+            computeStoreClient.computeAggregation().countGroupByValue(2, "location").execute(keySet).get();
```
Comment (haoxu07) [inline internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/storagenode/ReadComputeValidationTest.java]: We could define field name as private static String LOCATION_FIELD = "location" to improve the code readability.  Besides, we could pass "location" and "jobType" together to see if we can generate one response and then check the results to see if that is correct.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java (hunk 4 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java:51]: Amazing!

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/storagenode/ReadComputeValidationTest.java (hunk 6 above)
Comment (LeoLeo718) [reply internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/storagenode/ReadComputeValidationTest.java]: Yeah, you're right, that will improve readability

File: clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AvroComputeAggregationResponseTest.java (hunk 5 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AvroComputeAggregationResponseTest.java]: Awesome

**Review by haoxu07 — APPROVED** [review]
Overall LGTM! Maybe we could explore some unit test checking the type of casting error:

`T key = (T) value;`

**Comment by LeoLeo718** [comment]
> Overall LGTM! Maybe we could explore some unit test checking the type of casting error:
> 
> `T key = (T) value;`

Okay!

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    // Validate fields exist in schema
+    Schema valueSchema = schemaReader.getValueSchema(schemaReader.getLatestValueSchemaId());
+    for (String fieldName: fieldNames) {
+      if (fieldName == null) {
+        throw new VeniceClientException("Field name cannot be null");
+      }
+      if (fieldName.isEmpty()) {
+        throw new VeniceClientException("Field name cannot be empty");
+      }
+
```
Comment (majisourav99) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java]: once we check for 'if (fieldNames == null || fieldNames.length == 0)` these checks still needed?

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      Map<String, Integer> fieldTopKMap) {
+    this.computeResults = computeResults;
+    this.fieldTopKMap = fieldTopKMap;
+  }
+
+  @Override
+  public <T> Map<T, Integer> getValueToCount(String field) {
+    // Quick check: if field doesn't exist in fieldTopKMap, return empty map
+    if (!fieldTopKMap.containsKey(field)) {
+      return new LinkedHashMap<>();
```
Comment (majisourav99) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java]: return Collections.emptyMap() instead

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java (hunk 7 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationRequestBuilder.java]: Yeah, you're right, that's redundant

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java (hunk 8 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java]: wonderful

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+      valueToCount.merge(key, 1, Integer::sum);
+    }
+
+    // Sort by count in descending order
+    Map<T, Integer> sortedMap = new LinkedHashMap<>();
+    valueToCount.entrySet()
+        .stream()
+        .sorted(Map.Entry.<T, Integer>comparingByValue().reversed())
+        .limit(fieldTopKMap.get(field))
+        .forEach(entry -> sortedMap.put(entry.getKey(), entry.getValue()));
```
Comment (majisourav99) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java]: nit: you can do 
` .collect(
                  LinkedHashMap::new,
                  (m, e) -> m.put(e.getKey(), e.getValue()),
                  Map::putAll
              );`

Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroComputeAggregationResponse.java]: got it

**Comment by LeoLeo718** [comment]
> LGTM. left a minor comment



> LGTM. left a minor comment

Thank you!

**Review by majisourav99 — APPROVED** [review]
Thanks a lot Leo!
